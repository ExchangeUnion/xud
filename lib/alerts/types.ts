import { AlertType, ChannelSide } from '../constants/enums';

type BaseAlert = {
  type: AlertType;
  message: string;
  date: number;
};

export type BalanceAlertEvent = {
  /** The total balance of the channel when the alert is triggered. */
  totalBalance: number;
  /** The side of the balance either local or remote. */
  side: ChannelSide;
  /** The balance that triggered the alert. */
  sideBalance: number;
  /** The alert threshold in percentage, e.g. 10 means %10. */
  bound: number;
  /** The currency of the channel. */
  currency: string;
};
export type BalanceAlert = BaseAlert & BalanceAlertEvent;

export type Alert = BalanceAlert;
