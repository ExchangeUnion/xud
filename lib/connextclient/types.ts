/**
 * The configurable options for the connext client.
 */
export type ConnextClientConfig = {
  disable: boolean;
  host: string;
  port: number;
  webhookhost: string;
  webhookport: number;
};

/** General information about the state of this connext client. */
export type ConnextInfo = {
  status: string;
  error?: string;
  address?: string;
  chain?: string;
  version?: string;
};

/** Response for eth_blockNumber call  */
export type ConnextBlockNumberResponse = {
  result: string;
};

/** Response for eth_gasPrice call  */
export type EthproviderGasPriceResponse = {
  result: string;
};

type Balance = {
  amount: string[];
  to: string[];
};

/** Connext channel details */
export type ConnextChannelDetails = {
  assetIds: string[];
  balances: Balance[];
  channelAddress: string;
  merkleRoot: string;
  processedDepositsA: string[];
  processedDepositsB: string[];
  networkContext: {
    chainId: number;
    channelFactoryAddress: string;
    transferRegistryAddress: string;
    providerUrl: string;
  };
  nonce: number;
  alice: string;
  aliceIdentifier: string;
  bob: string;
  bobIdentifier: string;
  timeout: string;
  latestUpdate: {
    assetId: string;
    balance: Balance;
    channelAddress: string;
    details: {
      totalDepositsAlice: string;
      totalDepositsBob: string;
    };
    fromIdentifier: string;
    nonce: number;
    aliceSignature: string;
    bobSignature: string;
    toIdentifier: string;
    type: string;
  };
  defundNonce: string;
  inDispute: boolean;
};

/**
 * The payload for tokenPayment call.
 */
export type ConnextTransferRequest = {
  type: 'HashlockTransfer';
  channelAddress: string;
  amount: string;
  assetId: string;
  details: {
    lockHash: string;
    expiry: string;
  };
  recipient: string;
  meta: {
    routingId: string;
  };
  publicIdentifier: string;
};

/**
 * The response for errors.
 */
export type ConnextErrorResponse = { message: string };

export type ConnextConfig = {
  publicIdentifier: string;
  signerAddress: string;
  index: number;
};

/**
 * The response for /config call.
 */
export type ConnextConfigResponse = ConnextConfig[];

/**
 * The response for /channel call.
 */
export type ConnextChannelResponse = string[];

/**
 * The response for balance call.
 */
export type ConnextChannelBalanceResponse = {
  freeBalanceOffChain: string;
  nodeFreeBalanceOffChain: string;
  freeBalanceOnChain: string;
};

export type GetBlockByNumberResponse = {
  result: {
    difficulty: string;
    extraData: string;
    gasLimit: string;
    gasUsed: string;
    hash: string;
    logsBloom: string;
    miner: string;
    mixHash: string;
    nonce: string;
    number: string;
    parentHash: string;
    receiptsRoot: string;
    sha3Uncles: string;
    size: string;
    stateRoot: string;
    timestamp: string;
    totalDifficulty: string;
    transactions: string[];
    transactionsRoot: string;
    uncles: string[];
  };
};

/**
 * The response for hashLockTransfer call.
 */
export type ConnextTransferResponse = {
  channelAddress: string;
  transferId: string;
  routingId: string;
};

/**
 * The response for withdraw call.
 */
export type ConnextWithdrawResponse = {
  channelAddress: '0xa929dB0530daB525596f5d48Fb5C322fDa8A337B';
  transferId: '0xc2e4592d3fb6c02ee1d3b07bed83b5914f8ca084b0f91d6b80bf8107e58c9c38';
};

type ConnextRoutingPath = {
  recipient: string;
  recipientChainId: number;
  recipientAssetId: string;
};

export type ConnextTransfer = {
  channelFactoryAddress: string;
  assetId: string;
  chainId: number;
  channelAddress: string;
  balance: {
    amount: string[];
    to: string[];
  };
  initiator: string;
  responder: string;
  initialStateHash: string;
  transferDefinition: string;
  transferEncodings: string[];
  transferId: string;
  transferState: {
    lockHash: string;
    expiry: string;
  };
  transferTimeout: string;
  meta: {
    requireOnline: boolean;
    routingId: string;
    path: ConnextRoutingPath[];
    senderIdentifier?: string;
  };
  transferResolver?: {
    preImage?: string;
  };
};

export type TransfersByRoutingIdResponse = ConnextTransfer[];

export type ExpectedIncomingTransfer = {
  rHash: string;
  units: bigint;
  expiry: number;
  tokenAddress: string;
  routingId: string;
};

export type ConnextPreimageRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: string;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      requireOnline: boolean;
      routingId: string;
      path: [
        {
          recipient: string;
          recipientChainId: number;
          recipientAssetId: string;
        },
      ];
    };
    transferResolver: {
      preImage: string;
    };
  };
  conditionType: string;
};

export type ConnextIncomingTransferRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: number;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      routingId: string;
    };
  };
  conditionType: string;
};

export type ConnextDepositConfirmedRequest = {
  id: string;
  data?: {
    amount: {
      _hex: string;
    };
    assetId: string;
    hash: string;
  };
};

export type ProvidePreimageEvent = {
  rHash: string;
  preimage: string;
};

export type TransferReceivedEvent = {
  tokenAddress: string;
  rHash: string;
  expiry: number;
  units: bigint;
  routingId: string;
};
