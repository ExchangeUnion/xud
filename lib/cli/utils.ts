import colors from 'colors/safe';
import { accessSync, watch } from 'fs';
import os from 'os';
import path from 'path';

const SATOSHIS_PER_COIN = 10 ** 8;

export function getDefaultCertPath() {
  switch (os.platform()) {
    case 'win32': {
      const homeDir = process.env.LOCALAPPDATA!;
      return path.join(homeDir, 'Xud', 'tls.cert');
    }
    case 'darwin': {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
    default: {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
  }
}

export function shorten(key: string, length = 10) {
  if (key.length <= length * 2 + 3) {
    return key;
  }
  return `${key.slice(0, length)}...${key.slice(key.length - length)}`;
}

export function trim(key: string, length = 10) {
  if (key.length <= length + 3) {
    return '';
  }
  return `${key.slice(0, length)}...`;
}

export const generateHeaders = (headers: string[]) => {
  return headers.map((header) => {
    return colors.blue(header);
  });
};

/** Returns a number of coins as an integer number of satoshis. */
export const coinsToSats = (coinsQuantity: number) => {
  return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
};

/** Returns a number of satoshis as a string representation of coins with up to 8 decimal places. */
export const satsToCoinsStr = (satsQuantity: number) => {
  return (satsQuantity / SATOSHIS_PER_COIN).toFixed(8).replace(/\.?0+$/, '');
};

/** Waits up to 5 seconds for the tls.cert file to be created in case this is the first time xud has been run. */
export const waitForCert = (certPath: string) => {
  return new Promise<void>((resolve, reject) => {
    try {
      accessSync(certPath);
      resolve();
    } catch (err) {
      if (err.code === 'ENOENT') {
        const certDir = path.dirname(certPath);
        const certFilename = path.basename(certPath);
        const fsWatcher = watch(certDir, (event, filename) => {
          if (event === 'change' && filename === certFilename) {
            clearTimeout(timeout);
            fsWatcher.close();
            resolve();
          }
        });
        const timeout = setTimeout(() => {
          fsWatcher.close();
          reject(`timed out waiting for cert to be created at ${certPath}`);
        }, 5000);
      } else {
        // we handle errors due to file not existing, otherwise reject
        reject(err);
      }
    }
  });
};

export function showSeed(seedMnemonicList: string[]) {
  const WORDS_PER_ROW = 4;
  const numberedMnemonic = seedMnemonicList.map((value, index) => {
    return `${index >= 9 ? '' : ' '}${index + 1}. ${value.padEnd(10)}`;
  });
  console.log('----------------------BEGIN XUD SEED---------------------');
  for (let n = 0; n < seedMnemonicList.length / WORDS_PER_ROW; n += 1) {
    console.log(numberedMnemonic.slice(n * WORDS_PER_ROW, (n + 1) * WORDS_PER_ROW).join(' '));
  }
  console.log('-----------------------END XUD SEED----------------------\n');

  console.log(`
Please write down your 24 word mnemonic. It will allow you to recover your xud \
node key and on-chain funds for the initialized wallets listed above should you \
forget your password or lose your device. Off-chain funds in channels can NOT \
be recovered with it and must be backed up and recovered separately. Keep it \
somewhere safe, it is your ONLY backup in case of data loss.
`);
}
