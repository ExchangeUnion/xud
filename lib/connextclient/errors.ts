import errorCodesPrefix from '../constants/errorCodesPrefix';

const codesPrefix = errorCodesPrefix.CONNEXT;
const errorCodes = {
  CONNEXT_IS_DISABLED: codesPrefix.concat('.1'),
  INSUFFICIENT_BALANCE: codesPrefix.concat('.2'),
  TIMEOUT: codesPrefix.concat('.3'),
  INVALID: codesPrefix.concat('.4'),
  SERVER_ERROR: codesPrefix.concat('.5'),
  UNEXPECTED: codesPrefix.concat('.6'),
  INVALID_TOKEN_PAYMENT_RESPONSE: codesPrefix.concat('.9'),
  CONNEXT_HAS_NO_ACTIVE_CHANNELS: codesPrefix.concat('.11'),
  CONNEXT_CLIENT_NOT_INITIALIZED: codesPrefix.concat('.12'),
  CURRENCY_NOT_FOUND_BY_TOKENADDRESS: codesPrefix.concat('.13'),
  CURRENCY_MISSING: codesPrefix.concat('.14'),
  EXPIRY_MISSING: codesPrefix.concat('.15'),
  MISSING_SEED: codesPrefix.concat('.16'),
  INSUFFICIENT_COLLATERAL: codesPrefix.concat('.17'),
  NOT_FOUND: codesPrefix.concat('.18'),
  WITHDRAW_ADDRESS_MISSING: codesPrefix.concat('.19'),
};

const errors = {
  CONNEXT_IS_DISABLED: {
    message: 'connext is disabled',
    code: errorCodes.CONNEXT_IS_DISABLED,
  },
  CONNEXT_HAS_NO_ACTIVE_CHANNELS: {
    message: 'connext has no active channels',
    code: errorCodes.CONNEXT_HAS_NO_ACTIVE_CHANNELS,
  },
  CONNEXT_CLIENT_NOT_INITIALIZED: {
    message: 'connext client is not initialized',
    code: errorCodes.CONNEXT_CLIENT_NOT_INITIALIZED,
  },
  INSUFFICIENT_BALANCE: {
    message: 'insufficient balance to perform request',
    code: errorCodes.INSUFFICIENT_BALANCE,
  },
  TIMEOUT: {
    message: 'request timed out',
    code: errorCodes.TIMEOUT,
  },
  SERVER_ERROR: (statusCode: number, statusMessage?: string) => ({
    message: `connext server error ${statusCode}: ${statusMessage ?? ''}`,
    code: errorCodes.SERVER_ERROR,
  }),
  UNEXPECTED: (statusCode: number, statusMessage?: string) => ({
    message: `unexpected status from connext request ${statusCode}: ${statusMessage ?? ''}`,
    code: errorCodes.UNEXPECTED,
  }),
  NOT_FOUND: {
    message: 'connext returned not found response',
    code: errorCodes.NOT_FOUND,
  },
  WITHDRAW_ADDRESS_MISSING: {
    message: 'destination account for the withdrawal is missing',
    code: errorCodes.WITHDRAW_ADDRESS_MISSING,
  },
  INVALID_TOKEN_PAYMENT_RESPONSE: {
    message: 'connext TokenPaymentResponse is invalid',
    code: errorCodes.INVALID_TOKEN_PAYMENT_RESPONSE,
  },
  CURRENCY_NOT_FOUND_BY_TOKENADDRESS: (tokenAddress: string) => ({
    message: `could not find currency for tokenAddress ${tokenAddress}`,
    code: errorCodes.CURRENCY_NOT_FOUND_BY_TOKENADDRESS,
  }),
  CURRENCY_MISSING: {
    message: 'currency is missing',
    code: errorCodes.CURRENCY_MISSING,
  },
  EXPIRY_MISSING: {
    message: 'expiry is missing',
    code: errorCodes.EXPIRY_MISSING,
  },
  MISSING_SEED: {
    message: 'seed is missing',
    code: errorCodes.MISSING_SEED,
  },
  INSUFFICIENT_COLLATERAL: {
    message: 'channel collateralization in progress, please try again in ~1 minute',
    code: errorCodes.INSUFFICIENT_COLLATERAL,
  },
};

export { errorCodes };
export default errors;
