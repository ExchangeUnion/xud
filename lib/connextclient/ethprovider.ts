import { ethers } from 'ethers';
import { curry } from 'ramda';
import { from, Observable } from 'rxjs';
import { mergeMap } from 'rxjs/operators';
// This file will be a separate module with the above dependencies.

// gets the Ethereum provider object to read data from the chain
const getProvider = (host: string, port: number, chainId: number): ethers.providers.JsonRpcProvider => {
  return new ethers.providers.JsonRpcProvider({ url: `http://${host}:${port}/ethprovider/${chainId}` });
};

// gets the signer object necessary for write access (think unlock wallet)
const getSigner = (provider: ethers.providers.JsonRpcProvider, seed: string): ethers.Wallet => {
  return ethers.Wallet.fromMnemonic(seed).connect(provider);
};

// We curry getContract so that we can provide its arguments one at a time.
// This allows us to provide some of the necessary arguments (that we already have) before we export the function.
// Read more: https://ramdajs.com/docs/#curry
const getContract = curry(
  (signer: ethers.Wallet, contractAddress: string): ethers.Contract => {
    // we use the minimum viable contract ABI for ERC20 tokens
    // for full contract ABI we should compile it from the solidity source
    const erc20abi = ['function balanceOf(address) view returns (uint)', 'function transfer(address to, uint amount)'];
    return new ethers.Contract(contractAddress, erc20abi, signer);
  },
);

// Sends on-chain ERC20 transfer
// We also curry this function, just like the previous one.
// All the functions that we export out of the package will be curried
const onChainSendERC20 = curry(
  (
    signer: ethers.Wallet,
    contract: ethers.Contract,
    destinationAddress: string,
    units: string,
  ): Observable<ethers.ContractTransaction> => {
    // get the gas price
    return from(signer.provider.getGasPrice()).pipe(
      mergeMap(
        (gasPrice) =>
          // then send the transaction
          from(contract.transfer(destinationAddress, units, { gasPrice })) as Observable<ethers.ContractTransaction>,
      ),
    );
  },
);

// Sends on-chain ETH transfer
const onChainSendETH = curry(
  (signer: ethers.Wallet, destinationAddress: string, units: string): Observable<ethers.ContractTransaction> => {
    return from(signer.provider.getGasPrice()).pipe(
      mergeMap((gasPrice) => {
        const ether = ethers.utils.formatEther(units);
        const value = ethers.utils.parseEther(ether);
        return signer.sendTransaction({
          to: destinationAddress,
          value,
          gasPrice,
        });
      }),
    );
  },
);

// returns ETH on-chain balance for the address in wei
const getEthBalanceByAddress = curry((provider: ethers.providers.JsonRpcProvider, address: string) =>
  from(provider.getBalance(address)),
);

// returns ERC20 on-chain balance for the contract address in the smallest unit
const getERC20Balance = curry(
  (address: string, contract: ethers.Contract): Observable<ethers.BigNumber> => {
    return from(contract.balanceOf(address)) as Observable<ethers.BigNumber>;
  },
);

// This is the main function that has to be called before this package exposes more functions.
// Think of it as a constructor where we create the interal state of the module before
// we export more functionality to the consumer.
const getEthprovider = (host: string, port: number, chainId: number, seed: string) => {
  // create the internal state
  const provider = getProvider(host, port, chainId);
  const signer = getSigner(provider, seed);
  // because the functions below are curried we can only provide some of the arguments
  const getERC20BalanceWithSigner = getERC20Balance(signer.address);
  const getContractWithSigner = getContract(signer);
  const onChainSendERC20WithSigner = onChainSendERC20(signer);
  const onChainSendETHWithSigner = onChainSendETH(signer);
  const getEthBalanceByAddressWithProvider = getEthBalanceByAddress(provider);
  const onChainTransfer = (contractAddress: string, destinationAddress: string, units: string) => {
    if (contractAddress === ethers.constants.AddressZero) {
      return onChainSendETHWithSigner(destinationAddress, units);
    } else {
      const contract = getContractWithSigner(contractAddress);
      return onChainSendERC20WithSigner(contract, destinationAddress, units);
    }
  };
  // expose functionality to the consumer
  return {
    getEthBalance: () => from(signer.getBalance()),
    getEthBalanceByAddress: getEthBalanceByAddressWithProvider,
    getContract: getContractWithSigner,
    getERC20Balance: getERC20BalanceWithSigner,
    getERC20BalanceByAddress: getERC20Balance,
    onChainTransfer,
  };
};

type EthProvider = ReturnType<typeof getEthprovider>;

export { getEthprovider, EthProvider };
