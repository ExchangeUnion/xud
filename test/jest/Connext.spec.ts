import ConnextClient from '../../lib/connextclient/ConnextClient';
import errors from '../../lib/connextclient/errors';
import { SwapClientType } from '../../lib/constants/enums';
import { CurrencyInstance } from '../../lib/db/types';
import Logger from '../../lib/Logger';
import { PaymentState } from '../../lib/swaps/SwapClient';
import { getUnitConverter } from '../utils';
import { EthProvider } from '../../lib/connextclient/ethprovider';
import { of } from 'rxjs';

const MOCK_TX_HASH = '0x5544332211';
jest.mock('../../lib/utils/utils', () => {
  return {
    parseResponseBody: () => {
      return { txhash: MOCK_TX_HASH };
    },
  };
});
jest.mock('../../lib/Logger');
const mockedLogger = <jest.Mock<Logger>>(<any>Logger);

jest.mock('http', () => {
  return {
    request: jest.fn().mockImplementation((options, cb) => {
      if (options.path === '/deposit') {
        cb({ statusCode: 404 });
      }

      let errorCb: any;
      return {
        path: options.path,
        write: jest.fn(),
        end: jest.fn(),
        on: jest.fn().mockImplementation((event, cb) => {
          if (event === 'error') {
            errorCb = cb;
          }
        }),
        destroy: jest.fn().mockImplementation(() => {
          errorCb();
        }),
      };
    }),
  };
});

const ETH_ASSET_ID = '0x0000000000000000000000000000000000000000';
const USDT_ASSET_ID = '0xdAC17F958D2ee523a2206206994597C13D831ec7';
const XUC_ASSET_ID = '0x9999999999999999999999999999999999999999';

describe('ConnextClient', () => {
  let connext: ConnextClient;

  beforeEach(() => {
    const config = {
      disable: false,
      host: 'http://tester',
      port: 1337,
      webhookhost: 'http://testerson',
      webhookport: 7331,
    };
    const logger = new mockedLogger();
    logger.trace = jest.fn();
    logger.error = jest.fn();
    logger.debug = jest.fn();
    logger.warn = jest.fn();
    logger.info = jest.fn();
    const currencyInstances = [
      {
        id: 'ETH',
        tokenAddress: ETH_ASSET_ID,
        swapClient: SwapClientType.Connext,
      },
      {
        id: 'USDT',
        tokenAddress: USDT_ASSET_ID,
        swapClient: SwapClientType.Connext,
      },
      {
        id: 'XUC',
        tokenAddress: XUC_ASSET_ID,
        swapClient: SwapClientType.Connext,
      },
    ] as CurrencyInstance[];
    connext = new ConnextClient({
      config,
      currencyInstances,
      logger,
      unitConverter: getUnitConverter(),
      network: 'mainnet',
    });
  });

  describe('withdraw', () => {
    const MOCK_ETH_FREE_BALANCE_ON_CHAIN = 10n ** 18n; // 1 ETH
    const MOCK_USDT_FREE_BALANCE_ON_CHAIN = 10000n * 10n ** 6n; // 10000 USDT
    const DESTINATION_ADDRESS = '0x12345';
    const onChainTransfer = jest.fn(() => of(Promise.resolve({ hash: MOCK_TX_HASH })));

    beforeEach(() => {
      connext['getBalance'] = jest.fn().mockImplementation((currency: string) => {
        switch (currency) {
          case 'ETH':
            return { freeBalanceOnChain: MOCK_ETH_FREE_BALANCE_ON_CHAIN };
          case 'USDT':
            return { freeBalanceOnChain: MOCK_USDT_FREE_BALANCE_ON_CHAIN };
          default:
            return { freeBalanceOnChain: 0 };
        }
      });
      connext['ethProvider'] = ({
        onChainTransfer,
      } as unknown) as EthProvider;
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('fails with custom fee', async () => {
      expect.assertions(1);
      try {
        await connext.withdraw({
          currency: 'ETH',
          destination: DESTINATION_ADDRESS,
          amount: 123,
          fee: 1,
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it('fails to withdraw all ETH', async () => {
      expect.assertions(1);
      try {
        await connext.withdraw({
          currency: 'ETH',
          destination: DESTINATION_ADDRESS,
          all: true,
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it('fails when amount bigger than wallet balance', async () => {
      expect.assertions(1);
      try {
        await connext.withdraw({
          currency: 'ETH',
          destination: DESTINATION_ADDRESS,
          amount: 2 * 10 ** 8,
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it('withdraws all USDT', async () => {
      expect.assertions(3);
      const txhash = await connext.withdraw({
        currency: 'USDT',
        destination: DESTINATION_ADDRESS,
        all: true,
      });
      expect(onChainTransfer).toHaveBeenCalledTimes(1);
      expect(onChainTransfer).toHaveBeenCalledWith(USDT_ASSET_ID, DESTINATION_ADDRESS, MOCK_USDT_FREE_BALANCE_ON_CHAIN);
      expect(txhash).toEqual(MOCK_TX_HASH);
    });

    it('withdraws 5000 USDT amount', async () => {
      expect.assertions(3);
      const txhash = await connext.withdraw({
        currency: 'USDT',
        destination: DESTINATION_ADDRESS,
        amount: 5000 * 10 ** 8,
      });
      expect(onChainTransfer).toHaveBeenCalledTimes(1);
      expect(onChainTransfer).toHaveBeenCalledWith(USDT_ASSET_ID, DESTINATION_ADDRESS, '5000000000');
      expect(txhash).toEqual(MOCK_TX_HASH);
    });

    it('withdraws 0.000001 ETH amount', async () => {
      expect.assertions(3);
      const txhash = await connext.withdraw({
        currency: 'ETH',
        destination: DESTINATION_ADDRESS,
        amount: 1 * 10 ** 2,
      });
      expect(onChainTransfer).toHaveBeenCalledTimes(1);
      expect(onChainTransfer).toHaveBeenCalledWith(ETH_ASSET_ID, DESTINATION_ADDRESS, '1000000000000');
      expect(txhash).toEqual(MOCK_TX_HASH);
    });
  });

  describe('sendRequest', () => {
    it('deposit fails with 404', async () => {
      expect.assertions(1);
      try {
        await connext['sendRequest']('/deposit', 'POST', {
          assetId: ETH_ASSET_ID,
          amount: BigInt('100000').toString(),
        });
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  describe('lookupPayment', () => {
    it('returns PaymentState.Pending', async () => {
      expect.assertions(1);
      connext['getHashLockStatus'] = jest.fn().mockReturnValue({
        transferState: { expiry: '10001' },
        transferResolver: {},
      });
      connext['getHeight'] = jest.fn().mockReturnValue(10000);
      const result = await connext['lookupPayment']('0x12345', 'ETH');
      expect(result).toEqual({ state: PaymentState.Pending });
    });

    it('returns PaymentState.Completed with preimage', async () => {
      expect.assertions(1);
      connext['getHashLockStatus'] = jest.fn().mockReturnValue({
        transferState: { expiry: '10001' },
        transferResolver: { preImage: '0x1337' },
      });
      connext['getHeight'] = jest.fn().mockReturnValue(10000);
      const result = await connext['lookupPayment']('0x12345', 'ETH');
      expect(result).toEqual({
        state: PaymentState.Succeeded,
        preimage: '1337',
      });
    });

    it('returns PaymentState.Failed when preimage is hash zero', async () => {
      expect.assertions(1);
      connext['getHashLockStatus'] = jest.fn().mockReturnValue({
        transferState: { expiry: '10001' },
        transferResolver: { preImage: '0x0000000000000000000000000000000000000000000000000000000000000000' },
      });
      connext['getHeight'] = jest.fn().mockReturnValue(10000);
      const result = await connext['lookupPayment']('0x12345', 'ETH');
      expect(result).toEqual({ state: PaymentState.Failed });
    });

    it('returns PaymentState.Failed when EXPIRED', async () => {
      expect.assertions(1);
      connext['getHashLockStatus'] = jest.fn().mockReturnValue({
        transferState: { expiry: '10001' },
        transferResolver: {},
      });
      connext['getHeight'] = jest.fn().mockReturnValue(10001);
      connext['sendRequest'] = jest.fn().mockReturnValue(Promise.resolve());
      const hash = '8f28fb27a164ae992fb4808b11c137d06e8e7d9304043a6b7163323f7cf53920';
      const currency = 'ETH';
      const result = await connext['lookupPayment'](hash, currency);
      expect(result).toEqual({ state: PaymentState.Failed });
    });

    it('returns PaymentState.Pending when error is unknown', async () => {
      expect.assertions(1);
      connext['getHashLockStatus'] = jest.fn().mockImplementation(() => {
        throw new Error('unknown error');
      });
      const result = await connext['lookupPayment']('0x12345', 'ETH');
      expect(result).toEqual({ state: PaymentState.Pending });
    });

    it('returns PaymentState.Failed when error is NOT_FOUND', async () => {
      expect.assertions(1);
      connext['getHashLockStatus'] = jest.fn().mockImplementation(() => {
        throw errors.NOT_FOUND;
      });
      const result = await connext['lookupPayment']('0x12345', 'ETH');
      expect(result).toEqual({ state: PaymentState.Failed });
    });
  });

  describe('setReservedInboundAmount', () => {
    const amount = 50000000;
    const currency = 'ETH';

    beforeEach(() => {
      connext['sendRequest'] = jest.fn().mockResolvedValue(undefined);
    });

    it('requests collateral plus 3% buffer when we have none', async () => {
      connext['inboundAmounts'].set('ETH', 0);
      connext.setReservedInboundAmount(amount, currency);
      expect(connext['sendRequest']).toHaveBeenCalledWith(
        '/request-collateral',
        'POST',
        expect.objectContaining({
          assetId: ETH_ASSET_ID,
          amount: (amount * 1.03 * 10 ** 10).toLocaleString('fullwide', {
            useGrouping: false,
          }),
        }),
      );
    });

    it('requests collateral plus 3% buffer when we have some collateral already', async () => {
      connext['inboundAmounts'].set('ETH', amount * 0.5);
      connext.setReservedInboundAmount(amount, currency);
      expect(connext['sendRequest']).toHaveBeenCalledWith(
        '/request-collateral',
        'POST',
        expect.objectContaining({
          assetId: ETH_ASSET_ID,
          amount: (amount * 1.03 * 10 ** 10).toLocaleString('fullwide', {
            useGrouping: false,
          }),
        }),
      );
    });

    it('does not request collateral when we have more than enough to cover the reserved inbound amount', async () => {
      connext['inboundAmounts'].set('ETH', amount * 2);
      connext.setReservedInboundAmount(amount, currency);
      expect(connext['sendRequest']).toHaveBeenCalledTimes(0);
    });
  });

  describe('checkInboundCapacity', () => {
    const quantity = 20000000;
    const smallQuantity = 100;
    beforeEach(() => {
      connext['sendRequest'] = jest.fn().mockResolvedValue(undefined);
      connext['inboundAmounts'].set('ETH', 0);
    });

    it('requests collateral plus 5% buffer when there is none', async () => {
      expect(() => connext.checkInboundCapacity(quantity, 'ETH')).toThrowError(
        'channel collateralization in progress, please try again in ~1 minute',
      );

      expect(connext['sendRequest']).toHaveBeenCalledTimes(1);
      expect(connext['sendRequest']).toHaveBeenCalledWith(
        '/request-collateral',
        'POST',
        expect.objectContaining({
          assetId: ETH_ASSET_ID,
          amount: (quantity * 1.05 * 10 ** 10).toLocaleString('fullwide', {
            useGrouping: false,
          }),
        }),
      );
    });

    it('does not request collateral when there is a pending request', async () => {
      connext['requestCollateralPromises'].set('ETH', Promise.resolve());
      expect(() => connext.checkInboundCapacity(quantity, 'ETH')).toThrowError(
        'channel collateralization in progress, please try again in ~1 minute',
      );

      expect(connext['sendRequest']).toHaveBeenCalledTimes(0);
    });

    it('requests the full collateral amount even when there is some existing collateral', async () => {
      const partialCollateral = 5000;
      connext['inboundAmounts'].set('ETH', partialCollateral);

      expect(() => connext.checkInboundCapacity(quantity, 'ETH')).toThrowError(
        'channel collateralization in progress, please try again in ~1 minute',
      );

      expect(connext['sendRequest']).toHaveBeenCalledTimes(1);
      expect(connext['sendRequest']).toHaveBeenCalledWith(
        '/request-collateral',
        'POST',
        expect.objectContaining({
          assetId: ETH_ASSET_ID,
          amount: (quantity * 1.05 * 10 ** 10).toLocaleString('fullwide', {
            useGrouping: false,
          }),
        }),
      );
    });

    it('requests the hardcoded minimum if the collateral shortage is below it', async () => {
      const minCollateralRequestUnits = ConnextClient['MIN_COLLATERAL_REQUEST_SIZES']['ETH']! * 10 ** 10;

      expect(() => connext.checkInboundCapacity(smallQuantity, 'ETH')).toThrowError(
        'channel collateralization in progress, please try again in ~1 minute',
      );

      expect(connext['sendRequest']).toHaveBeenCalledTimes(1);
      expect(connext['sendRequest']).toHaveBeenCalledWith(
        '/request-collateral',
        'POST',
        expect.objectContaining({
          assetId: ETH_ASSET_ID,
          amount: minCollateralRequestUnits.toLocaleString('fullwide', {
            useGrouping: false,
          }),
        }),
      );
    });

    it('requests collateral plus 5% buffer for a small shortage when there is no hardcoded minimum for the currency', async () => {
      expect(() => connext.checkInboundCapacity(smallQuantity, 'XUC')).toThrowError(
        'channel collateralization in progress, please try again in ~1 minute',
      );

      expect(connext['sendRequest']).toHaveBeenCalledTimes(1);
      expect(connext['sendRequest']).toHaveBeenCalledWith(
        '/request-collateral',
        'POST',
        expect.objectContaining({
          assetId: XUC_ASSET_ID,
          amount: (smallQuantity * 1.05 * 10 ** 10).toLocaleString('fullwide', {
            useGrouping: false,
          }),
        }),
      );
    });

    it('does not request collateral or throw when there is sufficient collateral', async () => {
      connext['inboundAmounts'].set('ETH', quantity);
      connext.checkInboundCapacity(quantity, 'ETH');

      expect(connext['sendRequest']).toHaveBeenCalledTimes(0);
    });
  });

  describe('disconnect', () => {
    it('aborts pending requests, except critical ones', async () => {
      expect(connext['pendingRequests'].size).toEqual(0);

      connext['sendRequest'](connext['criticalRequestPaths'][0], '', {});
      connext['sendRequest']('/path1', '', {});
      connext['sendRequest']('/path1', '', {});
      connext['sendRequest']('/path2', '', {});
      connext['sendRequest'](connext['criticalRequestPaths'][1], '', {});
      expect(connext['pendingRequests'].size).toEqual(5);

      connext['disconnect']();
      expect(connext['pendingRequests'].size).toEqual(2);
    });
  });
});
