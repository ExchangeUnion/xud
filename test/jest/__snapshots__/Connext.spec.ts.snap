// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnextClient sendRequest deposit fails with 404 1`] = `
Object {
  "code": "8.18",
  "message": "connext returned not found response",
}
`;

exports[`ConnextClient withdraw fails to withdraw all ETH 1`] = `[Error: withdrawing all ETH is not supported yet]`;

exports[`ConnextClient withdraw fails when amount bigger than wallet balance 1`] = `
Object {
  "code": "8.2",
  "message": "insufficient balance to perform request",
}
`;

exports[`ConnextClient withdraw fails with custom fee 1`] = `[Error: setting fee for Ethereum withdrawals is not supported yet]`;
